apps/string_publisher/app.c

```diff
@@ -0,0 +1,83 @@
+#include <stdio.h>
+#include <unistd.h>
+
+#include <rcl/rcl.h>
+#include <rcl/error_handling.h>
+#include <std_msgs/msg/string.h>
+
+#include <rclc/rclc.h>
+#include <rclc/executor.h>
+
+#ifdef ESP_PLATFORM
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#endif
+
+#define RCCHECK(fn) { rcl_ret_t temp_rc = fn; if((temp_rc != RCL_RET_OK)){printf("Failed status on line %d: %d. Aborting.\n",__LINE__,(int)temp_rc);vTaskDelete(NULL);}}
+#define RCSOFTCHECK(fn) { rcl_ret_t temp_rc = fn; if((temp_rc != RCL_RET_OK)){printf("Failed status on line %d: %d. Continuing.\n",__LINE__,(int)temp_rc);}}
+
+//initialization
+rcl_publisher_t publisher;
+std_msgs__msg__String msg;
+int counter = 1;
+
+void timer_callback(rcl_timer_t * timer, int64_t last_call_time)
+{
+	RCLC_UNUSED(last_call_time);
+	if (timer != NULL) {
+		// assign message to publisher
+  		std_msgs__msg__String__init(&msg);
+  		const unsigned int PUB_MSG_CAPACITY = 30;
+  		msg.data.data = malloc(PUB_MSG_CAPACITY);
```

Comment:
Maybe you could initialize the string message and allocate its memory on `appMain`. This way you don't need to use all that allocations.

Also, use the `rcl_allocator_t allocator` instead of direct `malloc` or `free` calls

Reply:
Hi,
In the case of foxy, I will still need to free the memory even though I assign it inside the main function. The code will retain the variables until the code is running and will free the memory once the code is complete. 

For Galactic I will have to test the code as I have been working with foxy. Somehow on this website, ( https://micro.ros.org/docs/tutorials/advanced/handling_type_memory/ ) there is no mention of memory leak or freeing memory once it's been allocated. Can you check it once to confirm?

Reply:
Yes, but this way you will only allocate/free memory once which is a good practice. Your current code can lead to heap fragmentation .

> I will still need to free the memory even though I assign it inside the main function. The code will retain the variables until the code is running and will free the memory once the code is complete.

I don't understand the following part, what are you asking exactly? 
You are supposed to free all memory manually allocated.

> ... there is no mention of memory leak or freeing memory once it's been allocated.

Reply:
I mean that there is no reference of freeing memory in the tutorial section. So we should add it for developers on the webpage.

Reply:
I mean that there is no reference of freeing memory in the tutorial section. So we should add it for developers on the webpage.

Reply:
We include a memory destruction part for our custom memory handler (Galactic):
> Finally, a destruction function is also provided for messages allocated in dynamic memory: ...

But the developer is expected to know how to use `malloc` and `free`, leaving the tutorial to specific micro-ROS message initialization.

Reply:
I have updated the initialization of msg as you suggested.